Add Intel TCC tutorial (#83)
canonical/real-time-ubuntu-docs: docs/tutorial/intel-tcc/test-application.md

# The test application

The test application is named **rt_linux_tutorial**.

> Its architecture consist of a control thread that performs random pointer chasing to simulate a workload and measures wake up jitter, execution time and cache misses.

Reviewer: ```suggestion
Its architecture consists of a control thread that performs random pointer chasing to simulate a workload and measures wake up jitter, execution time and cache misses.
```

The application also features a statistics handler thread that processes and logs timing information.
The statistics are either output to the console via STDOUT or published to an MQTT broker.

```{figure} images/rt_linux_tutorial.svg
   :width: 100%
   :alt: High-level design of rt app

   High-level design of the test application
```

```{note}
The pointer chasing buffer size should exceed the L2 cache size of your processor to see an improvement by cache partitioning described later in this tutorial.
You can change the buffer size in `rt_linux_tutorial.c` by editing the `WORKLOAD_BUFFER_SIZE` definition.
```

## Build the docker image for the test application

The source code for the test application is available in the code archive you downloaded earlier.

> First make any changes to the C program code, like modifying the `WORKLOAD_BUFFER_SIZE`.

Reviewer: ```suggestion
First make any required changes to the C program code, like modifying the `WORKLOAD_BUFFER_SIZE`.
```

Then build the Docker image by running the following command in the directory containing `Dockerfile`:

```bash
cd tutorial-intel-tcc-code
docker build -t intel_tcc_tutorial_image .
```

To verify that the image was built successfully, list all Docker images:

```bash
docker images
```
You should see `intel_tcc_tutorial_image` listed among the images.

Run a new interactive container using this image:

```bash
docker run -it --privileged --rm --network docker-compose_stats intel_tcc_tutorial_image
```

Inside the container the `rt_linux_tutorial` application can be started:

> ```bash
> root@container:~# ./rt_linux_tutorial -h
> Usage: rt_linux_tutorial [OPTIONS]
> Options:
>   -h        Display this help message
>   -i <time> Set the cycle time of the control thread in microseconds - default is 500us
>   -s <0|1>  Set the output method for statistics (0 for stdout, 1 for MQTT to localhost) - default stdout
> ```

Reviewer: ````suggestion
```{terminal}
:user: root
:host: container
:dir: ~
:input: ./rt_linux_tutorial -h

Usage: rt_linux_tutorial [OPTIONS]
Options:
  -h        Display this help message
  -i <time> Set the cycle time of the control thread in microseconds - default is 500us
  -s <0|1>  Set the output method for statistics (0 for stdout, 1 for MQTT to localhost) - default stdout
```
````

See https://documentation.ubuntu.com/real-time/en/latest/how-to/isolate-workload-cpusets/#configure-runtime-cpu-isolation if you want to see the rendered example.

Run it with MQTT support.

```bash
./rt_linux_tutorial -i 1000 -s 1
```

Data should start appearing on the Grafana dashboard.

## Troubleshooting

> If you don't see any data on the Grafana dashboard, try the following to debug the data flow.

Reviewer: I get the intent of this section but this doesn't actually tell you what to do to debug / what are the expected or common problems.

```suggestion
If you don't see any data on the Grafana dashboard, check the data flow and resolve any errors displayed in the messages and/or logs.
```

### Subscribe to the Mosquitto broker
```
mosquitto_sub -t '#'
```
You should see a continuous stream of JSON data, that looks similar to this:
```json
{"exec_time":170852,"wakeup_jitter":6592,"cache_misses":0,"ipc":0.036748599261045456}
```

### Check Telegraf logs

> ```

Reviewer: ````suggestion

```
````

For consistency; blank line after a heading.

cd tutorial-intel-tcc-code/docker-compose/
docker compose logs -f telegraf
```
You should see the startup sequence of Telegraf, loading InfluxDB as output plugin and MQTT as input.
There should not be any errors in the log.

> ```

Reviewer: Nitpick but for consistency might be good to add a blank line before and after a code block.
Helps with readability too.
Same comment for other occurrences.

telegraf | I! Loaded outputs: influxdb
...
telegraf | I! [inputs.mqtt_consumer] Connected [tcp://mosquitto:1883]
```

### Check InfluxDB logs

> ```

Reviewer: ````suggestion

```
````
